libraries/ms-helper/src/persist.c

```diff
@@ -0,0 +1,158 @@
+#include "persist.h"
+#include <inttypes.h>
+#include "log.h"
+#include "soft_timer.h"
+// The persistance layer allocates one page of flash so we can erase the entire page when full.
+// To reduce the number of erases, we partition the page into a number of sections. Each section
+// is the size of the specified blob plus a header. The header is used to mark the section as valid
+// and record the blob's size.
+//
+// By default, all sections are invalid. Only one section should be valid at a time. At each write,
+// all previous sections are marked as invalid and the new section is considered valid.
+//
+// Section format: [ marker (u32) | size (u32) | blob (u32 * n) ]
+// We initially write to the size field and blob data, skipping the marker. We then consider the
+// default unwritten value of marker as "valid" (0xFFFFFFFF). To invalidate the section, we write
+// 0x0 to the marker field. Because we skipped it earlier, this write succeeds and does not require
+// erasing the page. If all sections are invalid, we erase the entire page.
+//
+// At init, the persistance layer attemps to load the blob with the data stored in flash.
+
+// Erased flash defaults to all 1's
+#define PERSIST_VALID_MARKER 0xFFFFFFFF
+#define PERSIST_INVALID_SIZE 0xFFFFFFFF
+#define PERSIST_INVALID_ADDR UINTPTR_MAX
+#define PERSIST_BASE_ADDR FLASH_PAGE_TO_ADDR(PERSIST_FLASH_PAGE)
+#define PERSIST_END_ADDR (PERSIST_BASE_ADDR + FLASH_PAGE_BYTES)
+
+typedef struct PersistHeader {
+  uint32_t marker;
```

Comment:
Hmm, my reasoning for only supporting writing words was that it was more consistent with the rest of the system. In addition, since we have 2kb per page, I don't think saving 32 bits is that important. Thoughts?

Reply:
Fair enough. I'm just careful about size savings nowadays :p